contrib/cmake/build3rdParty/build3rdParty.cpp

```diff
@@ -0,0 +1,273 @@
+
+
+#if (__cplusplus >= 201703L && (!defined(__GNUC__) || (__GNUC__*100+__GNUC_MINOR__) >= 800))
+#include <filesystem>
+namespace fs = std::filesystem;
+#define USE_FILESYSTEM
+#elif !defined(__MINGW32__) && !defined(__ANDROID__) && (!defined(__GNUC__) || (__GNUC__*100+__GNUC_MINOR__) >= 503)
+#define USE_FILESYSTEM
+#ifdef WIN32
+#include <filesystem>
+namespace fs = std::experimental::filesystem;
+#else
+#include <experimental/filesystem>
+namespace fs = std::experimental::filesystem;
+#endif
+#else
+//for mac
+#include <filesystem>
+namespace fs = std::__fs::filesystem;
+#endif
+
+
+
+bool build = false;
+bool setup = false;
+bool removeUnusedPorts = false;
+fs::path portsFile;
+fs::path sdkRootPath;
+string triplet;
+
+map<string, string> ports;
+
+fs::path initialDir = fs::current_path();
+fs::path vcpkgDir = fs::current_path() / "vcpkg";
+fs::path cloneDir = fs::current_path() / "vcpkg_clone";
+
+
+bool readCommandLine(int argc, char* argv[]);
+void execute(string command);
+
+int main(int argc, char* argv[])
+try
+{
+    if (readCommandLine(argc, argv))
+    {
+        if (setup)
+        {
+            if (!fs::is_directory("vcpkg"))
+            {
+                execute("git clone https://github.com/microsoft/vcpkg.git");
+                execute("git clone --progress -v vcpkg vcpkg_clone");
+                fs::current_path("vcpkg");
+                #ifdef WIN32
+                    execute(".\\bootstrap-vcpkg.bat -disableMetrics");
+                #else
+                    execute("./bootstrap-vcpkg.sh -disableMetrics");
+                #endif
+            }
+            else
+            {
+                fs::current_path("vcpkg");
+            }
+
+            if (fs::exists(sdkRootPath / "contrib" / "cmake" / "vcpkg_extra_triplets" / (triplet + ".cmake")))
+            {
+                if (fs::exists(vcpkgDir / "triplets" / (triplet + ".cmake")))
+                {
+                    fs::remove(vcpkgDir / "triplets" / (triplet + ".cmake"));
+                }
+                cout << "Copying triplet from SDK: " << triplet << endl;
+                fs::copy(sdkRootPath / "contrib" / "cmake" / "vcpkg_extra_triplets" / (triplet+".cmake"), vcpkgDir / "triplets" / (triplet + ".cmake"));
+            }
+            else if (!fs::exists(vcpkgDir / "triplets" / (triplet + ".cmake")))
+            {
+                cout << "triplet not found in the SDK or in vcpkg: " << triplet << endl;
+                exit(1);
+            }
+
+            for (auto portPair : ports)
+            {
+                const string& portname = portPair.first;
+                const string& portversion = portPair.second;
+
+                if (fs::is_directory(fs::path("ports") / portname))
+                {
+                    cout << "Removing " << (vcpkgDir / "ports" / portname).u8string() << endl;
+                    fs::remove_all(vcpkgDir / "ports" / portname);
+                }
+                if (portversion.size() == 40)
```

Comment:
40  strikes me as a magic number, maybe could define it as a `constexpr` var or add a comment to explain that this branch (as I understand it) means "looks like a commit SHA?"

Reply:
yes exactly - a quick and dirty check for a commit SHA.   Feel free to improve it, thanks